### PR TITLE
8304928: Optimize ClassDesc.resolveConstantDesc

### PR DESCRIPTION
**Pull request by liach**
This patch optimizes ClassDesc.resolveConstantDesc for array classes. Otherwise, the performance of reference class resolution remain unchanged.

A benchmark comparing resolution for reference arrays, primitive arrays, and class or interfaces is included in this patch, and the result comparison on my machine is available at https://jmh.morethan.io/?gists=21c462f4879f6166b3308a41c092d51e,1aeae6ff40a129f89dbfedcb874cfd37

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304928](https://bugs.openjdk.org/browse/JDK-8304928): Optimize ClassDesc.resolveConstantDesc

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13252/head:pull/13252` \
`$ git checkout pull/13252`

Update a local copy of the PR: \
`$ git checkout pull/13252` \
`$ git pull https://git.openjdk.org/jdk.git pull/13252/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13252`

View PR using the GUI difftool: \
`$ git pr show -t 13252`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13252.diff">https://git.openjdk.org/jdk/pull/13252.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/constant/ConstantUtils.java [hunk 1]
```diff
@@ -126,6 +126,10 @@ static String dropFirstAndLastChar(String s) {
         return s.substring(1, s.length() - 1);
     }
 
+    static boolean isPrimitiveArray(String referenceTypeDescriptorString) {
+        return referenceTypeDescriptorString.charAt(referenceTypeDescriptorString.length() - 1) != ';';
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/constant/ConstantUtils.java:130]: it also needs to check if it starts with `[`


```suggestion
        return referenceTypeDescriptorString.startsWith("[") &&  
                  referenceTypeDescriptorString.charAt(referenceTypeDescriptorString.length() - 1) != ';';
```

File: test/micro/org/openjdk/bench/java/lang/constant/ReferenceClassDescResolve.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                    return clazz;
+                }
+            }
+        },
+        NEW {
+            @Override
+            Class<?> call(MethodHandles.Lookup lookup, ClassDesc cd) throws ReflectiveOperationException {
+                return resolveReference(lookup, cd);
+            }
+        },
```
Comment (mlchung) [inline test/micro/org/openjdk/bench/java/lang/constant/ReferenceClassDescResolve.java:120]: I think this benchmark doesn't need this case as that's the implementation of `resolveConstantDesc`.   Was it there when you developed this fix?   Same for MethodTypeDesc benchmark.

File: test/micro/org/openjdk/bench/java/lang/constant/MethodTypeDescResolve.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * Compare array clone with equivalent System.arraycopy-based routines
+ */
+@BenchmarkMode(Mode.Throughput)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 3, time = 2)
+@Measurement(iterations = 6, time = 1)
+@Fork(1)
+@State(Scope.Benchmark)
+public class MethodTypeDescResolve {
```
Comment (mlchung) [inline test/micro/org/openjdk/bench/java/lang/constant/MethodTypeDescResolve.java:53]: Better to separate this benchmark from this PR.   Probably the optimization done in [PR 10991](https://github.com/openjdk/jdk/pull/10991) should be a separate patch from JDK-8294961 and this benchmark could be included there.

File: src/java.base/share/classes/java/lang/constant/ConstantUtils.java (hunk 1 above)
Comment (liach) [reply src/java.base/share/classes/java/lang/constant/ConstantUtils.java:130]: I intend this to be a `ReferenceClassDesc`-specifc check. I think moving `isPrimitiveArray` as a private method in `ReferenceClassDesc` would be better, as the `startsWith("[")` is already implied (at least before valhalla, non-array, i.e. class or interface descriptors must end with a semicolon.

**Comment by liach** [comment]
Done. Now the benchmark explicitly measures the 3 reference types, with new result in line with previous observations here: https://jmh.morethan.io/?gists=21c462f4879f6166b3308a41c092d51e,1aeae6ff40a129f89dbfedcb874cfd37

The `isPrimitiveArray` check is converted to a private method in `ReferenceClassDesc` as it applies only to reference type descriptor strings. Also included the rationale for not checking initial `[` in a comment.

**Comment by liach** [comment]
If you don't mind, please also check out #12986, which avoids casting the result to `Class` after calling `ClassDesc::resolveConstantDesc`.

**Comment by mlchung** [comment]
> The `isPrimitiveArray` check is converted to a private method in `ReferenceClassDesc` as it applies only to reference type descriptor strings. Also included the rationale for not checking initial `[` in a comment.

This would work.

**Review by mlchung — APPROVED** [review]
Looks good.  The new benchmark is cleaner.
